### PR TITLE
Decouple API Manager and API Microgateway

### PR DESCRIPTION
**Pull request by AmaliMatharaarachchi**

### CODE REVIEW

File: components/micro-gateway-core/src/main/ballerina/gateway/services/throttle_event_listener.bal [hunk 1]
```diff
@@ -19,56 +19,82 @@ import ballerina/log;
...
+                GlobalThrottleStreamDTO globalThrottleStreamDtoTM = {
+                    throttleKey: <string>m[THROTTLE_KEY],
+                    isThrottled: <boolean>m[IS_THROTTLED],
+                    expiryTimeStamp: <int>m[EXPIRY_TIMESTAMP] };
 
                 if (globalThrottleStreamDtoTM.isThrottled == true) {
                     putThrottleData(globalThrottleStreamDtoTM);
-                } else {
+                }
+                else {
```
Comment (Arshardh) [inline components/micro-gateway-core/src/main/ballerina/gateway/services/throttle_event_listener.bal]: Is there any reason to break the else statement to two lines, I noticed this in several places.

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/config/TOMLConfigParser.java [hunk 2]
```diff
@@ -81,4 +81,5 @@ public static void write(String configFilePath, Object content) throws ConfigPar
     private static <T> T parseString(String configFileContent, Class<T> type) {
         return new Toml().read(configFileContent).to(type);
     }
+
```
Comment (Arshardh) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/config/TOMLConfigParser.java]: shall we remove this blank line

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/constants/RESTServiceConstants.java [hunk 3]
```diff
@@ -37,18 +37,22 @@
...
+
+    public static final String DEFAULT_HOST = "https://localhost:9443";
     public static final String DEFAULT_TRUSTSTORE_PATH = "lib/platform/bre/security/ballerinaTruststore.p12";
     public static final String DEFAULT_TRUSTSTORE_PASS = "ballerina";
     public static final String DEFAULT_KEYSTORE_PATH = "lib/platform/bre/security/ballerinaKeystore.p12";
     public static final String DEFAULT_KEYSTORE_PASS = "ballerina";
 
     public static final String CERTIFICATE_ALIAS = "Alias";
     public static final String CERTIFICATE_TIER = "Tier";
+
```
Comment (Arshardh) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/constants/RESTServiceConstants.java]: shall we remove this blank line

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java [hunk 4]
```diff
@@ -74,93 +73,70 @@
 /**
  * This class represents the "setup" command and it holds arguments and flags specified by the user.
  */
-@Parameters(commandNames = "setup", commandDescription = "setup information")
-public class SetupCmd implements GatewayLauncherCmd {
+@Parameters(commandNames = "setup", commandDescription = "setup information") public class SetupCmd
```
Comment (Rajith90) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: Formatting issue

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java [hunk 5]
```diff
@@ -74,93 +73,70 @@
...
+        implements GatewayLauncherCmd {
+
     private static final Logger logger = LoggerFactory.getLogger(SetupCmd.class);
     private static PrintStream outStream = System.out;
 
-    @SuppressWarnings("unused")
-    @Parameter(hidden = true, required = true)
-    private List<String> mainArgs;
+    @SuppressWarnings("unused") @Parameter(hidden = true, required = true) private List<String> mainArgs;
```
Comment (Rajith90) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: Formatting issue

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java [hunk 6]
```diff
@@ -74,93 +73,70 @@
...
 
-    @SuppressWarnings("unused")
-    @Parameter(names = "--java.debug", hidden = true)
-    private String javaDebugPort;
+    @SuppressWarnings("unused") @Parameter(names = "--java.debug", hidden = true) private String javaDebugPort;
```
Comment (Rajith90) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: Check all place for formatting issues

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java [hunk 7]
```diff
@@ -442,26 +402,17 @@ public void execute() {
...
-                        .setPublisherEndpoint(publisherEndpoint)
-                        .setAdminEndpoint(adminEndpoint)
-                        .setRegistrationEndpoint(registrationEndpoint)
-                        .setTokenEndpoint(tokenEndpoint)
-                        .setUsername(username)
-                        .setClientId(clientID)
-                        .setClientSecret(encryptedCS)
-                        .setTrustStoreLocation(trustStoreLocation)
-                        .setTrustStorePassword(encryptedTrustStorePass)
+                Token token = configTokenValues.setUsername(username).setClientId(clientID).setClientSecret(encryptedCS)
```
Comment (Rajith90) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: Line length exceeding 100

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java [hunk 8]
```diff
@@ -482,25 +433,22 @@ public void execute() {
...
      */
     private void validateAPIGetRequestParams(String label, String apiName, String version) {
-        if ((StringUtils.isEmpty(label) && (StringUtils.isEmpty(apiName) || StringUtils.isEmpty(version))) ||
-                StringUtils.isNotEmpty(label) && (StringUtils.isNotEmpty(apiName) || StringUtils.isNotEmpty(version)) ||
-                (StringUtils.isEmpty(apiName) && StringUtils.isNotEmpty(version)) ||
-                (StringUtils.isNotEmpty(apiName) && StringUtils.isEmpty(version))) {
+        if ((StringUtils.isEmpty(label) && (StringUtils.isEmpty(apiName) || StringUtils.isEmpty(version)))
+                || StringUtils.isNotEmpty(label) && (StringUtils.isNotEmpty(apiName) || StringUtils.isNotEmpty(version))
+                || (StringUtils.isEmpty(apiName) && StringUtils.isNotEmpty(version)) || (StringUtils.isNotEmpty(apiName)
+                && StringUtils.isEmpty(version))) {
```
Comment (Rajith90) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: Check for line length

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/rest/APICorsConfigurationDTO.java [hunk 9]
```diff
@@ -25,6 +26,7 @@
 /**
  * CORS configuration for the APIDetailedDTO\n
  **/
+@JsonIgnoreProperties(ignoreUnknown = true)
```
Comment (Rajith90) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/rest/APICorsConfigurationDTO.java]: Add a comment to use of this property

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java [hunk 10]
```diff
@@ -556,30 +491,172 @@ private static void init(String projectName, String configPath, String deploymen
...
+        configTokenValues.setTokenEndpoint(tokenEndpoint);
+        configTokenValues.setRestVersion(restVersion);
+        configTokenValues.setBaseURL(token.getBaseURL());
+
+        isEndPointsNeeded = StringUtils.isEmpty(publisherEndpoint) || StringUtils.isEmpty(adminEndpoint) || StringUtils
+                .isEmpty(registrationEndpoint) || StringUtils.isEmpty(tokenEndpoint);
+
+        isBaseURLNeeded =
+                publisherEndpoint.contains("{baseURL}") || adminEndpoint.contains("{baseURL}") || registrationEndpoint
+                        .contains("{baseURL}") || tokenEndpoint.contains("{baseURL}") || isEndPointsNeeded;
```
Comment (Rajith90) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: Use a constant

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java [hunk 11]
```diff
@@ -556,30 +491,172 @@ private static void init(String projectName, String configPath, String deploymen
...
+
+        isRestVersionNeeded = publisherEndpoint.contains("{restVersion}") || adminEndpoint.contains("{restVersion}")
+                || registrationEndpoint.contains("{restVersion}") || isEndPointsNeeded;
```
Comment (Rajith90) [inline components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: use constant for {restVersion}

File: components/micro-gateway-core/src/main/ballerina/gateway/services/throttle_event_listener.bal (hunk 1 above)
Comment (AmaliMatharaarachchi) [reply components/micro-gateway-core/src/main/ballerina/gateway/services/throttle_event_listener.bal]: Fixed in https://github.com/wso2/product-microgateway/pull/332/commits/7dec453bb0eb0d526d5b95004e51f7b012efb678

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/config/TOMLConfigParser.java (hunk 2 above)
Comment (AmaliMatharaarachchi) [reply components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/config/TOMLConfigParser.java]: Fixed in https://github.com/wso2/product-microgateway/pull/332/commits/7dec453bb0eb0d526d5b95004e51f7b012efb678

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/constants/RESTServiceConstants.java (hunk 3 above)
Comment (AmaliMatharaarachchi) [reply components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/constants/RESTServiceConstants.java]: Fixed in https://github.com/wso2/product-microgateway/pull/332/commits/7dec453bb0eb0d526d5b95004e51f7b012efb678

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java (hunk 5 above)
Comment (AmaliMatharaarachchi) [reply components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: Fixed in https://github.com/wso2/product-microgateway/pull/332/commits/7dec453bb0eb0d526d5b95004e51f7b012efb678

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java (hunk 4 above)
Comment (AmaliMatharaarachchi) [reply components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: Fixed in https://github.com/wso2/product-microgateway/pull/332/commits/7dec453bb0eb0d526d5b95004e51f7b012efb678

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java (hunk 7 above)
Comment (AmaliMatharaarachchi) [reply components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: Fixed in https://github.com/wso2/product-microgateway/pull/332/commits/7dec453bb0eb0d526d5b95004e51f7b012efb678

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java (hunk 6 above)
Comment (AmaliMatharaarachchi) [reply components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: Fixed in https://github.com/wso2/product-microgateway/pull/332/commits/7dec453bb0eb0d526d5b95004e51f7b012efb678

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java (hunk 8 above)
Comment (AmaliMatharaarachchi) [reply components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: Fixed in https://github.com/wso2/product-microgateway/pull/332/commits/7dec453bb0eb0d526d5b95004e51f7b012efb678

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java (hunk 10 above)
Comment (AmaliMatharaarachchi) [reply components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: Fixed in https://github.com/wso2/product-microgateway/pull/332/commits/7dec453bb0eb0d526d5b95004e51f7b012efb678

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/rest/APICorsConfigurationDTO.java (hunk 9 above)
Comment (AmaliMatharaarachchi) [reply components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/model/rest/APICorsConfigurationDTO.java]: Fixed in https://github.com/wso2/product-microgateway/pull/332/commits/7dec453bb0eb0d526d5b95004e51f7b012efb678

File: components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java (hunk 11 above)
Comment (AmaliMatharaarachchi) [reply components/micro-gateway-cli/src/main/java/org/wso2/apimgt/gateway/cli/cmd/SetupCmd.java]: Fixed in https://github.com/wso2/product-microgateway/pull/332/commits/7dec453bb0eb0d526d5b95004e51f7b012efb678

**Pull request merged**